File: shared/plugin-common/src/main/scala/org/bitlap/skywalking/apm/plugin/common/AbstractThreadingPoolInterceptor.java
package org.bitlap.skywalking.apm.plugin.common;

import org.apache.skywalking.apm.agent.core.context.ContextManager;
import org.apache.skywalking.apm.agent.core.context.ContextSnapshot;
import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;

import java.lang.reflect.Method;

public abstract class AbstractThreadingPoolInterceptor implements InstanceMethodsAroundInterceptor {
    @Override
    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
        if (!ContextManager.isActive()) {
            return;
        }

        if (allArguments == null || allArguments.length < 1) {
            return;
        }


        Object argument = allArguments[0];

        // Avoid duplicate enhancement, such as the case where it has already been enhanced by RunnableWrapper or CallableWrapper with toolkit.
        if (argument instanceof EnhancedInstance && ((EnhancedInstance) argument).getSkyWalkingDynamicField() instanceof ContextSnapshot) {
            return;
        }


        Object wrappedObject = wrap(argument);
        if (wrappedObject != null) {
            allArguments[0] = wrappedObject;
        }
    }

    /**
     * wrap the Callable or Runnable object if needed
     *
     * @param param Callable or Runnable object
     * @return Wrapped object or null if not needed to wrap
     */
    public abstract Object wrap(Object param);

    @Override
    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Object ret) throws Throwable {
        return ret;
    }

    @Override
    public void handleMethodException(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, Throwable t) {
        AgentUtils$.MODULE$.logError(t);
    }
}

File: shared/plugin-common/src/main/scala/org/bitlap/skywalking/apm/plugin/common/AbstractThreadingPoolInterceptor.scala
package org.bitlap.skywalking.apm.plugin.common

import java.lang.reflect.Method

import org.apache.skywalking.apm.agent.core.context.*
import org.apache.skywalking.apm.agent.core.logging.api.LogManager
import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.*

abstract class AbstractThreadingPoolInterceptor extends InstanceMethodsAroundInterceptor {

  private lazy val LOGGER = LogManager.getLogger(classOf[AbstractThreadingPoolInterceptor])

  override def beforeMethod(
    objInst: EnhancedInstance,
    method: Method,
    allArguments: Array[Any],
    argumentsTypes: Array[Class[?]],
    result: MethodInterceptResult
  ): Unit = {
    LOGGER.info(s"ThreadingPool name: ${objInst.getClass.getName}, method: ${method.getName}")
    if !ContextManager.isActive then return
    if allArguments == null || allArguments.length < 1 then return
    val argument = allArguments(0)
    // Avoid duplicate enhancement, such as the case where it has already been enhanced by RunnableWrapper or CallableWrapper with toolkit.
    argument match
      case instance: EnhancedInstance if instance.getSkyWalkingDynamicField.isInstanceOf[ContextSnapshot] => return
      case _                                                                                              =>
    val wrappedObject = wrap(argument, objInst.getClass.getName, method.getName)
    if wrappedObject != null then allArguments.update(0, wrappedObject)
  }

  def wrap(param: Any, className: String, methodName: String): Any

  override def afterMethod(
    objInst: EnhancedInstance,
    method: Method,
    allArguments: Array[Any],
    argumentsTypes: Array[Class[?]],
    ret: Object
  ): AnyRef = ret

  override def handleMethodException(
    objInst: EnhancedInstance,
    method: Method,
    allArguments: Array[Any],
    argumentsTypes: Array[Class[?]],
    t: Throwable
  ): Unit =
    if ContextManager.isActive then ContextManager.activeSpan.log(t)
}
